### PR TITLE
Clarify usage of "arguments" and "parameters" in error messages

### PR DESCRIPTION
**Pull request by HertzDevil**
See #10374.

### CODE REVIEW

File: src/compiler/crystal/semantic/lib.cr [hunk 1]
```diff
@@ -57,11 +57,11 @@ class Crystal::Call
     named_args.each do |named_arg|
       found_index = external.args.index { |arg| arg.name == named_arg.name }
       unless found_index
-        named_arg.raise "no argument named '#{named_arg.name}'"
+        named_arg.raise "no parameter named '#{named_arg.name}'"
       end
 
       if covered[found_index]
-        named_arg.raise "argument '#{named_arg.name}' already specified"
+        named_arg.raise "argument for parameter '#{named_arg.name}' already specified"
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/lib.cr:64]: Is "for parameter" really helpful to improve this message?

File: src/compiler/crystal/types.cr [hunk 2]
```diff
@@ -2918,7 +2913,7 @@ module Crystal
     def instantiate(type_vars)
       types = type_vars.map do |type_var|
         unless type_var.is_a?(Type)
-          type_var.raise "argument to Proc must be a type, not #{type_var}"
+          type_var.raise "argument to Union must be a type, not #{type_var}"
```
Comment (bcardiff) [inline src/compiler/crystal/types.cr:2916]: If this error message was wrong it seems we are lacking some spec, right? Can it be added in this PR?

File: src/compiler/crystal/types.cr [hunk 3]
```diff
@@ -939,21 +937,18 @@ module Crystal
...
-
+      check_macro_param_count(a_macro, args_size)
       hooks = @hooks ||= [] of Hook
       hooks << Hook.new(kind, a_macro)
     end
 
+    private def check_macro_param_count(a_macro, expected_count)
+      param_count = a_macro.args.size
+      if param_count != expected_count
+        raise TypeException.new "wrong number of parameters for macro '#{a_macro.name}' (given #{param_count}, expected #{expected_count})"
```
Comment (bcardiff) [inline src/compiler/crystal/types.cr:948]: We are being more terse here in case of expected_count <= 1. But it's fine.

**Review by bcardiff — APPROVED** [review]

File: src/compiler/crystal/types.cr (hunk 2 above)
Comment (bcardiff) [reply src/compiler/crystal/types.cr:2916]: Thanks!

File: src/compiler/crystal/semantic/lib.cr (hunk 1 above)
Comment (HertzDevil) [reply src/compiler/crystal/semantic/lib.cr:64]: Parameters always have a name but this isn't the case for arguments (in some other error messages they are referenced by numbers like `#1` and so on).

Though using "parameter ... already specified" here is also fine to me.

Comment (straight-shoota) [reply src/compiler/crystal/semantic/lib.cr:64]: But this is about a call, so it's a duplicate argument, not a duplicate parameter.

Comment (HertzDevil) [reply src/compiler/crystal/semantic/lib.cr:64]: There cannot be duplicate arguments. Positional arguments are always identifiable by unique indices and duplicate named arguments are detected by the parser.

Comment (straight-shoota) [reply src/compiler/crystal/semantic/lib.cr:64]: Technically this is correct, yes. But from a user's perspective I would prefer not to be that specific here.
`Argument 'foo' already specified` is a completely sufficient error message.

Comment (HertzDevil) [reply src/compiler/crystal/semantic/lib.cr:64]: Technical correctness is the whole point of this issue. I don't see why this particular error message would be one that users prefer to be less specific than the other messages involving arguments.

Comment (straight-shoota) [reply src/compiler/crystal/semantic/lib.cr:64]: Why would shortening `argument` to `argument for parameter` be less correct? It's just more concise. Would the shortened form cause any ambiguity or misunderstanding of the error message's meaning?

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**